helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -109,19 +109,22 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
       return Optional.empty();
     }
 
-    Function<AssignableNode, Float> calculatePoints =
+    Function<AssignableNode, Double> calculatePoints =
         (candidateNode) -> getAssignmentNormalizedScore(candidateNode, replica, clusterContext);
 
     return candidateNodes.stream().max(Comparator.comparing(calculatePoints));
   }
 
-  private float getAssignmentNormalizedScore(AssignableNode node, AssignableReplica replica,
+  private double getAssignmentNormalizedScore(AssignableNode node, AssignableReplica replica,
       ClusterContext clusterContext) {
-    float sum = 0;
+    double sum = 0;
     for (Map.Entry<SoftConstraint, Float> softConstraintEntry : _softConstraints.entrySet()) {
       SoftConstraint softConstraint = softConstraintEntry.getKey();
       float weight = softConstraintEntry.getValue();
-      sum += weight * softConstraint.getAssignmentNormalizedScore(node, replica, clusterContext);
+      if (weight != 0) {
```

Comment:
Is it necessary? Weight comes from our manual setting and it will never be 0 otherwise it makes no sense to have the corresponding soft constraint implementation

Reply:
For that statement, I am not so sure. For example, I have a cluster don't care about top state assignment. So I set the weight of top state constraint to be zero. Then Helix does not need to do that calculation.